src/js/background/sync.js

```diff
@@ -0,0 +1,483 @@
+    const dataIsReliable = await sync.storageArea.dataIsReliable();
+    if (hasSyncStorage && dataIsReliable) await restore();
+
+    await sync.storageArea.backup();
+    return;  
+  },
+
+  async addContextualIdentityListeners(listenerList) {
+    if(!listenerList) listenerList = syncCIListenerList;
+    await browser.contextualIdentities.onCreated.addListener(listenerList[0]);
+    await browser.contextualIdentities.onRemoved.addListener(listenerList[1]);
+    await browser.contextualIdentities.onUpdated.addListener(listenerList[2]);
+  },
+
+  async removeContextualIdentityListeners(listenerList) {
+    if(!listenerList) listenerList = syncCIListenerList;
+    await browser.contextualIdentities.onCreated.removeListener(listenerList[0]);
+    await browser.contextualIdentities.onRemoved.removeListener(listenerList[1]);
+    await browser.contextualIdentities.onUpdated.removeListener(listenerList[2]);
+  },
+
+  async hasContextualIdentityListeners(listenerList) {
+    if(!listenerList) listenerList = syncCIListenerList;
+    return (
+      await browser.contextualIdentities.onCreated.hasListener(listenerList[0]) &&
+      await browser.contextualIdentities.onRemoved.hasListener(listenerList[1]) &&
+      await browser.contextualIdentities.onUpdated.hasListener(listenerList[2])
+    );
+  }
+
+};
+
+sync.init();
+
+async function restore() {
+  console.log("restore");
+  await reconcileIdentities();
+  await reconcileSiteAssignments();
+  return;
+}
+
+/*
+ * Checks for the container name. If it exists, they are assumed to be the
+ * same container, and the color and icon are overwritten from sync, if
+ * different.
+ */
+async function reconcileIdentities(){
+  console.log("reconcileIdentities");
+
+  // first delete any from the deleted list
+  const deletedIdentityList =
+    await sync.storageArea.getDeletedIdentityList();
+  // first remove any deleted identities
+  for (const deletedUUID of deletedIdentityList) {
+    const deletedCookieStoreId = 
+      await identityState.lookupCookieStoreId(deletedUUID);
+    if (deletedCookieStoreId){
+      try{
+        await browser.contextualIdentities.remove(deletedCookieStoreId);
+      } catch (error) {
+        // if the identity we are deleting is not there, that's fine.
+        console.error("Error deleting contextualIdentity", deletedCookieStoreId);
+        continue;
+      }
+    }
+  }
+
+  const localIdentities = await browser.contextualIdentities.query({});
+  const syncIdentities = 
+    await sync.storageArea.getIdentities();
+  const cookieStoreIDmap = 
+    await sync.storageArea.getCookieStoreIDMap();
+  // now compare all containers for matching names.
+  for (const syncIdentity of syncIdentities) {
+    syncIdentity.macAddonUUID = cookieStoreIDmap[syncIdentity.cookieStoreId];
+    if (syncIdentity.macAddonUUID){
+      const localMatch = localIdentities.find(
+        localIdentity => localIdentity.name === syncIdentity.name
+      );
```

Comment:
This seems to cover my question already, but just to be sure: How will it be handled if the MAC sync update gets pushed out to users that have multiple devices? Does that effectively duplicate containers when the update arrives before the sync storage is updated or does this specific name check prevent that?

Reply:
This name check should prevent that. It will dupe them if the names are not identical, however, and it will just pick one for the site assignment, with preference to whichever container got into sync first.

Reply:
Sounds good, thanks!